hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheRecordStore.java

```diff
@@ -164,64 +162,73 @@ public AbstractCacheRecordStore(String cacheNameWithPrefix, int partitionId, Nod
+        this.valueComparator = getValueComparatorOf(cacheConfig.getInMemoryFormat());
+        this.clearExpiredRecordsTask = cacheService.getExpirationManager().getTask();
 
-    private void initializeStatisticsAndFactories(String cacheNameWithPrefix) {
         if (cacheConfig.isStatisticsEnabled()) {
             statistics = cacheService.createCacheStatIfAbsent(cacheNameWithPrefix);
         }
-        if (cacheConfig.getCacheLoaderFactory() != null) {
-            Factory<CacheLoader> cacheLoaderFactory = cacheConfig.getCacheLoaderFactory();
-            cacheLoaderFactory = injectDependencies(cacheLoaderFactory);
-            cacheLoader = cacheLoaderFactory.create();
-            cacheLoader = injectDependencies(cacheLoader);
-        }
-        if (cacheConfig.getCacheWriterFactory() != null) {
-            Factory<CacheWriter> cacheWriterFactory = cacheConfig.getCacheWriterFactory();
-            cacheWriterFactory = injectDependencies(cacheWriterFactory);
-            cacheWriter = cacheWriterFactory.create();
-            cacheWriter = injectDependencies(cacheWriter);
-        }
+        injectDependencies(evictionPolicyEvaluator.getEvictionPolicyComparator());
+        TenantControl tenantControl = nodeEngine
+                .getTenantControlService()
+                .getTenantControl(ICacheService.SERVICE_NAME, cacheNameWithPrefix);
+        cacheLoader = TenantContextual.create(this::initCacheLoader,
+                () -> cacheConfig.getCacheLoaderFactory() != null, tenantControl);
+        cacheWriter = TenantContextual.create(this::initCacheWriter,
+                () -> cacheConfig.getCacheWriterFactory() != null, tenantControl);
+        defaultExpiryPolicy = TenantContextual.create(this::initDefaultExpiryPolicy,
+                this::defaultExpiryPolicyExists, tenantControl);
+        init();
+    }
+
+    private CacheLoader initCacheLoader() {
+        Factory<CacheLoader> cacheLoaderFactory = cacheConfig.getCacheLoaderFactory();
+        injectDependencies(cacheLoaderFactory);
+        CacheLoader loader = cacheLoaderFactory.create();
+        injectDependencies(loader);
+        registerResourceIfItIsClosable(loader);
+        return loader;
+    }
+
+    private CacheWriter initCacheWriter() {
+        Factory<CacheWriter> cacheWriterFactory = cacheConfig.getCacheWriterFactory();
+        injectDependencies(cacheWriterFactory);
+        CacheWriter writer = cacheWriterFactory.create();
+        injectDependencies(writer);
+        registerResourceIfItIsClosable(writer);
+        return writer;
+    }
+
+    private ExpiryPolicy initDefaultExpiryPolicy() {
+        Factory<ExpiryPolicy> expiryPolicyFactory = cacheConfig.getExpiryPolicyFactory();
+        injectDependencies(expiryPolicyFactory);
+        ExpiryPolicy expiryPolicy = expiryPolicyFactory.create();
+        injectDependencies(expiryPolicy);
+        registerResourceIfItIsClosable(defaultExpiryPolicy);
```

Comment:
should read `registerResourceIfItIsClosable(expiryPolicy)`, at this point the `defaultExpiryPolicy` is yet unset.

Reply:
Addressed in a separate commit, together with tenant control for IMap, IList, ISet and IQueue. PTAL.